image-builder/image_builder.sh

```diff
@@ -76,55 +76,83 @@ OK "init is installed"
 OK "Agent installed"
 [ "$(id -u)" -eq 0 ] || die "$0: must be run as root"
 
+ROOTFS_SIZE=$(du -B 1MB -s "${ROOTFS}" | awk '{print $1}')
 BLOCK_SIZE=${BLOCK_SIZE:-4096}
-IMG_SIZE=${IMG_SIZE:-80}
-
-info "Creating raw disk with size ${IMG_SIZE}M"
-qemu-img create -q -f raw "${IMAGE}" "${IMG_SIZE}M"
-OK "Image file created"
-
-# Kata runtime expect an image with just one partition
-# The partition is the rootfs content
-
-info "Creating partitions"
-parted "${IMAGE}" --script "mklabel gpt" \
-"mkpart ext4 1M -1M"
-OK "Partitions created"
 
-# Get the loop device bound to the image file (requires /dev mounted in the
-# image build system and root privileges)
-DEVICE=$(losetup -P -f --show "${IMAGE}")
+# Extra buffer space for the raw disk in percentage.
+IMG_BUFFER=30
 
-#Refresh partition table
-partprobe "${DEVICE}"
+# Calculate image size based on the rootfs
+calculate_img_size()
+{
+    IMG_SIZE=${IMG_SIZE:-$(((($ROOTFS_SIZE*$IMG_BUFFER)/100)+$ROOTFS_SIZE))}
+}
 
-MOUNT_DIR=$(mktemp -d osbuilder-mount-dir.XXXX)
-info "Formating Image using ext4 format"
-mkfs.ext4 -q -F -b "${BLOCK_SIZE}" "${DEVICE}p1"
-OK "Image formated"
+# Cleanup
+cleanup()
+{
+    sync
+    umount -l ${MOUNT_DIR}
+    fsck -D -y "${DEVICE}p1"
+    losetup -d "${DEVICE}"
+}
 
-info "Mounting root paratition"
-mount "${DEVICE}p1" "${MOUNT_DIR}"
-OK "root paratition mounted"
+create_rootfs()
+{
+    calculate_img_size
```

Comment:
Can add and extra validation to break the recursion:
- IMG_SIZE is not negative 
- IMG_SIZE is not bigger than MAX_IMAGE_SIZE(  2G? ) 

To make more the recursion depends on the image size can it receive image_size as parameter 
local image_size=$1
if (( image_size > MAX_SIZE )) ;  break/fail
if (( image_size <  MIN_SIZE )) ; break/fail

Reply:
Done. 

```
./image-builder/image_builder.sh -s 2049  "${PWD}/fedora_rootfs"
ERROR: Image size should not be greater than 2048 MB.
```
```
./image-builder/image_builder.sh -s -2049  "${PWD}/fedora_rootfs"
ERROR: Image size has to be greater than 0 MB.
```